File: c++/Sort/QuickSort.cpp
#include<iostream>
using namespace std;

void QuickSort(int s[], int l, int r)  
{  
    if (l< r)  
    {        
        int i = l, j = r, x = s[l];  
        while (i < j)  
        {  
            while(i < j && s[j]>= x) // 从右向左找第一个小于x的数  
                j--;   
            if(i < j)
                s[i++] = s[j];  
            while(i < j && s[i]< x) // 从左向右找第一个大于等于x的数  
                i++;   
            if(i < j)  
                s[j--] = s[i];  
        }  
        s[i] = x;  
        QuickSort(s, l, i - 1); // 递归调用  
        QuickSort(s, i + 1, r);  
    }  
}  

int main(){
    int arr[10] = {1,2,3,4,5,22,3,7,1,2};
    for( int i=0 ; i<10 ; i++ ){
        cout<<arr[i]<<" ";
    }
    cout<<endl;
    
    QuickSort( arr , 0 , 9 );
    for( int i=0 ; i<10 ; i++ ){
        cout<<arr[i]<<" ";
    }
    cout<<endl;
    return 0;
}

File: go/Sort/QuickSort.go
package main

import (
	"fmt"
	"math/rand"
)

func QuickSort(arr []int) []int {

	if len(arr) <= 1 {
		return arr
	}

	median := arr[rand.Intn(len(arr))]

	low_part := make([]int, 0, len(arr))
	high_part := make([]int, 0, len(arr))
	middle_part := make([]int, 0, len(arr))

	for _, item := range arr {
		switch {
		case item < median:
			low_part = append(low_part, item)
		case item == median:
			middle_part = append(middle_part, item)
		case item > median:
			high_part = append(high_part, item)
		}
	}

	low_part = QuickSort(low_part)
	high_part = QuickSort(high_part)

	low_part = append(low_part, middle_part...)
	low_part = append(low_part, high_part...)

	return low_part
}

func main() {
	arr := [10]int{1, 4, 8, 7, 6, 5, 9, 10, 2, 3}
	fmt.Println("Source:", arr)

	QuickSort(arr)

	fmt.Println("Sort:  ", arr)

}
